Added handlers in sympy.assumptions

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Continued https://github.com/sympy/sympy/pull/14465

#### Brief description of what is fixed or changed
Examples to the docstrings of transcendental,
infinite, and commutative have been added
in sympy.assumptions.ask

#### Other comments
Thanks to @oscargus for his review at https://github.com/sympy/sympy/pull/14465.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * Examples in sympy.assumptions.ask have been added.
<!-- END RELEASE NOTES -->

gdp1807 committed:
Added examples in sympy.assumptions.ask

Examples to the docstrings of transcendental,
infinite, and commutative have been added
in sympy.assumptions.ask

oscargus commented:
Thanks! Some unclear issues though... Not sure if all are related to your documentation though. The ones where nothing is returned are not obvious.

czgdp1807 commented on sympy/assumptions/ask.py:
```diff
@@ -241,8 +241,19 @@ def transcendental(self):
+        ========
+        >>> from sympy import ask, Q, pi, exp
+        >>> ask(Q.transcendental(exp(1)))
+        >>>
```
Two ways, either `ask(Q.transcendental(exp(1)))` or we can use `print(ask(Q.transcendental(exp(1))))` and then show `None` in the next line. I will go with the first one.

oscargus replied:
Either way, I do not think it will show True below? Maybe that is the problem then?

oscargus replied:
In general I guess it is better to avoid `print`, but I guess if the purpose is to show `None` it can be clarifying.

czgdp1807 replied:
I think it's an issue with `Predicate`. `e` is a transcendental number. I will fix this in this PR.

czgdp1807 committed:
Update ask.py

divyanshu132 commented on sympy/assumptions/ask.py:
```diff
@@ -706,8 +728,20 @@ def commutative(self):
+        >>> from sympy import ask, Q, MatrixSymbol, I
+        >>> X=MatrixSymbol('X', 2, 2)
+        >>> ask(Q.commutative(X))
+        >>>
```
may be this can be removed.

czgdp1807 replied:
There are some issues with the output. I am working on it.

czgdp1807 committed:
Added AskTranscendentalHandler in assumptions

AskTranscendentalHandler class have been added to
assumptions.handlers.calculus for handling transcendental numbers.

czgdp1807 committed:
Merge branch 'assumptions_docs' of https://github.com/czgdp1807/sympy into assumptions_docs

czgdp1807 commented:
@oscargus @asmeurer Please review the changes and additions I have made to the PR.

oscargus commented on sympy/assumptions/ask.py:
```diff
@@ -706,8 +726,19 @@ def commutative(self):
+        Examples
+        ========
+        >>> from sympy import ask, Q, MatrixSymbol, I
+        >>> X=MatrixSymbol('X', 2, 2)
```
Spaces around =.

oscargus commented on sympy/assumptions/ask.py:
```diff
@@ -706,8 +726,19 @@ def commutative(self):
+        ========
+        >>> from sympy import ask, Q, MatrixSymbol, I
+        >>> X=MatrixSymbol('X', 2, 2)
+        >>> ask(Q.commutative(X))
```
I think this is one of the tests that are failing.

oscargus commented on sympy/assumptions/handlers/calculus.py:
```diff
@@ -229,3 +229,130 @@ def log(expr, assumptions):
+
+    Examples of usage:
+
+    >>> from sympy import Symbol, Q, E, Pi
```
pi

oscargus commented on sympy/assumptions/handlers/calculus.py:
```diff
@@ -229,3 +229,130 @@ def log(expr, assumptions):
+    >>> from sympy.assumptions.handlers.calculus import AskTranscendentalHandler
+    >>> from sympy.abc import x
+    >>> a = AskTranscendentalHandler()
+    >>> a.Add(E + Pi)
```
pi

oscargus commented:
I have no idea how the assumption system works, but can at least comment on what I this k is wrong with the tests...

czgdp1807 committed:
Added AskInfiniteHandler in assumptions

AskInfiniteHandler has been added in
assumptions.handlers.calculus. Few bugs
were also removed and code reusability has
been enhanced.

czgdp1807 commented:
@oscargus Can you suggest something to reduce repetition of code or current implementation is fine for `AskInfiniteHandlers`.?

czgdp1807 committed:
Removed trivial bugs in docs

czgdp1807 commented:
@asmeurer I think, `AskCommutativeHandler` cannot handle matrices. I think there should be something like, `MatAdd` and `MatMul` or some changes should be made to `Basic` in `AskCommutativeHandler` for matrices. 

Example Code
```
from sympy import ask, Q, MatrixSymbol, I
from sympy.abc import x, y
X = MatrixSymbol('X', 2, 2)
Y = MatrixSymbol('Y', 2, 2)
print(ask(Q.commutative(X+Y)))
```

Associated Errors
```
Traceback (most recent call last):
  File "/home/gagandeep/sympy_debug.py", line 110, in <module>
    print(ask(Q.commutative(X+Y)))
  File "/home/gagandeep/sympy/sympy/assumptions/ask.py", line 1303, in ask
    res = key(expr)._eval_ask(assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 110, in _eval_ask
    return self.func.eval(self.arg, assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 193, in eval
    res = eval(expr, assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/handlers/common.py", line 49, in Basic
    if not ask(Q.commutative(arg), assumptions):
  File "/home/gagandeep/sympy/sympy/assumptions/ask.py", line 1303, in ask
    res = key(expr)._eval_ask(assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 110, in _eval_ask
    return self.func.eval(self.arg, assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 193, in eval
    res = eval(expr, assumptions)
  File "/home/gagandeep/sympy/sympy/assumptions/handlers/common.py", line 49, in Basic
    if not ask(Q.commutative(arg), assumptions):
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 166, in __call__
    return AppliedPredicate(self, expr)
  File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 66, in __new__
    arg = _sympify(arg)
  File "/home/gagandeep/sympy/sympy/core/sympify.py", line 405, in _sympify
    return sympify(a, strict=True)
  File "/home/gagandeep/sympy/sympy/core/sympify.py", line 327, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: 'X'

```
After some debugging the reason for this error is absence of `MatAdd` or `MatrixSymbol` and hence it redirects to `AskCommutativeHandler.Basic` resulting in the above error.

czgdp1807 committed:
Removed a bug in docs

czgdp1807 commented:
> @asmeurer I think, `AskCommutativeHandler` cannot handle matrices. I think there should be something like, `MatAdd` and `MatMul` or some changes should be made to `Basic` in `AskCommutativeHandler` for matrices.
> 
> Example Code
> 
> ```
> from sympy import ask, Q, MatrixSymbol, I
> from sympy.abc import x, y
> X = MatrixSymbol('X', 2, 2)
> Y = MatrixSymbol('Y', 2, 2)
> print(ask(Q.commutative(X+Y)))
> ```
> Associated Errors
> 
> ```
> Traceback (most recent call last):
>   File "/home/gagandeep/sympy_debug.py", line 110, in <module>
>     print(ask(Q.commutative(X+Y)))
>   File "/home/gagandeep/sympy/sympy/assumptions/ask.py", line 1303, in ask
>     res = key(expr)._eval_ask(assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 110, in _eval_ask
>     return self.func.eval(self.arg, assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 193, in eval
>     res = eval(expr, assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/handlers/common.py", line 49, in Basic
>     if not ask(Q.commutative(arg), assumptions):
>   File "/home/gagandeep/sympy/sympy/assumptions/ask.py", line 1303, in ask
>     res = key(expr)._eval_ask(assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 110, in _eval_ask
>     return self.func.eval(self.arg, assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 193, in eval
>     res = eval(expr, assumptions)
>   File "/home/gagandeep/sympy/sympy/assumptions/handlers/common.py", line 49, in Basic
>     if not ask(Q.commutative(arg), assumptions):
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 166, in __call__
>     return AppliedPredicate(self, expr)
>   File "/home/gagandeep/sympy/sympy/assumptions/assume.py", line 66, in __new__
>     arg = _sympify(arg)
>   File "/home/gagandeep/sympy/sympy/core/sympify.py", line 405, in _sympify
>     return sympify(a, strict=True)
>   File "/home/gagandeep/sympy/sympy/core/sympify.py", line 327, in sympify
>     raise SympifyError(a)
> sympy.core.sympify.SympifyError: SympifyError: 'X'
> ```
> After some debugging I discovered that the reason for this error is absence of `MatAdd` or `MatrixSymbol` and hence it redirects to `AskCommutativeHandler.Basic` resulting in the above error.

@jksuom @oscarbenjamin Any views on this from your side?

jksuom commented:
`MatrixSymbol` is a `Basic` object whose arguments are not `Basic` objects. (It is skipped in `test_args`.) It might be possible to change its construction in the same way as `Symbol` is handled: `name` is an attribute that is not included in `args`.

czgdp1807 commented:
I am working on your suggestions. As far as I can understand, I think to resolve the errors related to the above code, `MatrixSymbol` is needed to be restructured.

czgdp1807 commented:
I added `__slots__ = ['name']` in `MatrixSymbol`. Due to this one of the test is failing. The details are given below,

```
Traceback (most recent call last):
  File "/home/gagandeep/sympy/sympy/matrices/expressions/tests/test_blockmatrix.py", line 29, in test_bc_transpose
    assert bc_transpose(Transpose(BlockMatrix([[A, B], [C, D]]))) == \
  File "/home/gagandeep/sympy/sympy/matrices/expressions/blockmatrix.py", line 373, in bc_transpose
    return BlockMatrix(block_collapse(expr.arg).blocks.applyfunc(transpose).T)
  File "/home/gagandeep/sympy/sympy/matrices/expressions/blockmatrix.py", line 297, in block_collapse
    result = rule(expr)
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 11, in exhaustive_rl
    new, old = rule(expr), expr
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 44, in chain_rl
    expr = rule(expr)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 15, in <lambda>
    return chain(lambda expr: sall(bottom_up(rule, fns), fns)(expr), rule)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 37, in all_rl
    return new(op(expr), *args)
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 44, in chain_rl
    expr = rule(expr)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 15, in <lambda>
    return chain(lambda expr: sall(bottom_up(rule, fns), fns)(expr), rule)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 37, in all_rl
    return new(op(expr), *args)
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 44, in chain_rl
    expr = rule(expr)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 15, in <lambda>
    return chain(lambda expr: sall(bottom_up(rule, fns), fns)(expr), rule)
  File "/home/gagandeep/sympy/sympy/strategies/traverse.py", line 37, in all_rl
    return new(op(expr), *args)
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 44, in chain_rl
    expr = rule(expr)
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 11, in exhaustive_rl
    new, old = rule(expr), expr
  File "/home/gagandeep/sympy/sympy/strategies/core.py", line 32, in conditioned_rl
    if cond(expr):
  File "/home/gagandeep/sympy/sympy/matrices/expressions/blockmatrix.py", line 288, in <lambda>
    hasbm = lambda expr: isinstance(expr, MatrixExpr) and expr.has(BlockMatrix)
  File "/home/gagandeep/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/gagandeep/sympy/sympy/core/basic.py", line 126, in __hash__
    h = hash((type(self).__name__,) + self._hashable_content())
  File "/home/gagandeep/sympy/sympy/matrices/expressions/matexpr.py", line 664, in _hashable_content
    return (self.name, self.shape)
AttributeError: name

```
I don't know why `AttributeError` is being thrown, even though `name` is already defined as an attribute.
Please help me with this.

jksuom commented:
I think that more is needed than adding `name` to `__slots__`. It would probably be best to sympify the `args` of a `MatrixSymbol`. One way to do that is to make `args[0]` a Symbol whose name is the actual name string. That is what was proposed in https://github.com/sympy/sympy/pull/15121. Then it might not be necessary to add new handlers.

czgdp1807 commented:
sympifying name works well and there is no need to add new handlers for matrices. However, due to this some tests are failing and I think @asmeurer is working on it at #15121. So, I think I should wait for that PR to be merged. What do you say? In addition, if there any reviews on the handlers I have added then please provide them. :-) Thanks for helping me out @jksuom.

jksuom commented:
Perhaps you could also pull locally https://github.com/sympy/sympy/pull/15121 and test the examples that you wanted to add.

czgdp1807 commented:
@asmeurer Please provide your comments/reviews on this PR, since it is related to #15121 that you are working on.

czgdp1807 commented:
@asmeurer Waiting for reply on my [previous comment](https://github.com/sympy/sympy/pull/15902#issuecomment-477893003).

czgdp1807 commented:
@oscarbenjamin Please update the labels to `sympy's turn`.

czgdp1807 commented:
@asmeurer @smichr What do you say? Is this still needed? If it is then I can continue it in a new PR. Thanks.

czgdp1807 commented:
Closing this PR.
